File: getAllDrinks/getAllDrinks.js
// This was a script to download the complete list of drinks. This was run with nodejs and is not something that I expect you to know, but I'm providing it, if you're curious!
const fs = require('fs');

all_drinks = JSON.parse(fs.readFileSync("cocktails.json", { encoding: "utf8" }))

fetch("https://www.thecocktaildb.com/api/json/v1/1/list.php?i=list")
    .then(res => res.json())
    .then(res => {
        all_ingredients = res.drinks.map(i => i.strIngredient1)
        downloadByIngredients(all_ingredients);
    })

function downloadByIngredients(all_ingredients) {
    const fetchPromisesArray = all_ingredients.map(i => {
        return fetch("https://www.thecocktaildb.com/api/json/v1/1/filter.php?i=" + i)
            .then(res => res.json());
    });

    Promise.all(fetchPromisesArray)
        .then(allResults => {
            // allResults = allResults.drinks.flat();
            // allResults = allResults.reduce((acc, v) => v.drinks ? acc.concat(v.drinks) : acc, []);
            // require('fs').writeFileSync("allCocktails.json", JSON.stringify(allResults));
            allResults.drinks.forEach(drink => addDrink(drink));
        })
        .then(_ => all_drinks.sort((a, b) => a.strName < b.strName))
        .then(_ => fs.writeFileSync("new_cocktails.json", JSON.stringify(all_drinks)))
}

function addDrink(new_drink) {
    if (!all_drinks.some(drink => drink.strName === new_drink.strName))
        all_drinks.push(all_drinks);
}

File: getAllDrinks/getAllDrinks.py
import requests
import json
import time

with open("cocktails.json", encoding="utf-8") as f:
    all_drinks = json.load(f)

all_drinks = [drink for drink in all_drinks if drink]  # Remove nulls

# First get all the ingredients
all_ingredients = requests.get(
    "https://www.thecocktaildb.com/api/json/v1/1/list.php?i=list"
    ).json()
all_ingredients = all_ingredients["drinks"]

# For each ingredient, get matching drinks
for curr_ingredient in all_ingredients:
    print("INGRE:", curr_ingredient["strIngredient1"])
    new_drinks = requests.get(
        "https://www.thecocktaildb.com/api/json/v1/1/filter.php?i=" + curr_ingredient["strIngredient1"]
        ).json()
    new_drinks = new_drinks["drinks"]

    # Matching drinks only have id, name, and pic. Need to get details
    for curr_drink in new_drinks:
        # First, check to see if it doesn't already exist in all_drinks
        if not any(drink for drink in all_drinks if drink["strDrink"] == curr_drink["strDrink"]):
            drink_detail = requests.get(
                "https://www.thecocktaildb.com/api/json/v1/1/lookup.php?i=" + curr_drink["idDrink"]
                ).json()
            drink_detail = drink_detail["drinks"][0]
            if drink_detail:
                all_drinks.append(drink_detail)
    
    # The API is rate-limited. Let's slow things down
    time.sleep(0.1)

# Write to temp file, in case the sort crashes...
with open("unsorted_cocktails.json", "w") as f:
    json.dump(all_drinks, f)

all_drinks.sort(key=lambda d: d["strDrink"])

with open("new_cocktails.json", "w") as f:
    json.dump(all_drinks, f)
